operator/buffer/memory.go

```diff
@@ -29,36 +31,135 @@ func NewMemoryBufferConfig() *MemoryBufferConfig {
+	// an indefinite block or longer blocks to maxChunkDelay.
+	readLock sync.Mutex
 }
 
 // Add adds an entry onto the buffer.
 // Is a blocking call if the buffer is full
 func (m *MemoryBuffer) Add(ctx context.Context, e *entry.Entry) error {
-	return nil
+	// If buffer is closed don't allow this operation
+	if m.closed {
+		return ErrBufferedClosed
+	}
+
+	// Insert into buffer or error if context finishes before we can
+	select {
+	case <-ctx.Done():
+		return fmt.Errorf("ctx error adding to buffer: %w", ctx.Err())
+	case m.buf <- e:
+		return nil
+	}
 }
 
 // Read reads from the buffer.
 // Read will block until the there are MaxChunkSize entries or we have block as long as MachChunkDelay.
 func (m *MemoryBuffer) Read(ctx context.Context) ([]*entry.Entry, error) {
-	return nil, nil
+	// If buffer is closed don't allow this operation
+	if m.closed {
+		return nil, ErrBufferedClosed
+	}
+
+	m.readLock.Lock()
+	defer m.readLock.Unlock()
+
+	entries := make([]*entry.Entry, 0, m.maxChunkSize)
+
+	// Time for max Chunk delay
+	// Using a timer here rather than a context with deadline as we don't want to confuse a context timeout with a maxChunkDelay timeout
+	timer := time.NewTimer(m.maxChunkDelay)
+	defer timer.Stop()
+
+LOOP:
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, fmt.Errorf("ctx error during buffer read: %w", ctx.Err())
+		case <-timer.C:
+			// Timer has reached maxChunkDelay break and return
+			break LOOP
+		case e := <-m.buf:
+			entries = append(entries, e)
+
+			// If we've reached the maxChunkSize break and return
+			if len(entries) == int(m.maxChunkSize) {
+				break LOOP
+			}
+		}
+	}
+
+	return entries, nil
 }
 
-// Drain drains all contents currently in the buffer to the returned entry
+// Drain drains all contents currently in the buffer at the time of calling Drain.
 func (m *MemoryBuffer) Drain(ctx context.Context) ([]*entry.Entry, error) {
-	return nil, nil
+	// If buffer is closed don't allow this operation
+	if m.closed {
+		return nil, ErrBufferedClosed
+	}
+
+	m.readLock.Lock()
+	defer m.readLock.Unlock()
+
+	// Get the current size of the buffer
+	bufSize := len(m.buf)
+
+	entries := make([]*entry.Entry, bufSize)
+
+	// Loop until we hit the bufSize or the context cancels
+	for i := 0; i < bufSize; i++ {
+		select {
+		case <-ctx.Done():
+			return nil, fmt.Errorf("ctx error during buffer drain: %w", ctx.Err())
+		case e := <-m.buf:
+			entries[i] = e
+		}
+	}
+
+	return entries, nil
 }
 
 // Close runs cleanup code for buffer
 func (m *MemoryBuffer) Close() error {
+	// Acquire lock so we can't close while Read or Drain is occuring.
+	// It will also protect against multiple Close being called at once
+	m.readLock.Lock()
+	defer m.readLock.Unlock()
+
+	// Buffer already closed
+	if m.closed {
+		return nil
+	}
+
+	// Mark as closed so any operations after this point won't execute
+	m.closed = true
+
+	// Close the buffer channel then drain it
+	close(m.buf)
+	for range m.buf {
+	}
```

Comment:
Never seen this pattern before, haha. 

I'm kinda wondering how this + drain would work. It seems like we assume that writers to the memory buffer won't write anymore at some point, then `Drain()` is called, then `Close()`?

Reply:
I guess in my head I thought Close would close it for writing, but drain could still be called on it; But thinking about it, maybe that causes cleanup issue.

Reply:
Good questions so once you close a buffer (regardless of the implementation) I wouldn't think it's usable any more. Calling `close(m.buf)` will cause means nothing can push to the channel. 

You can call `Drain` whenever you want and as many times as you want. Calling `Close` should mean the buffer is done with and cleaned up.